File: sdks/ios/Sources/XMTPiOS/Codecs/EditMessageCodec.swift
import Foundation

public let ContentTypeEditMessageRequest = ContentTypeID(
	authorityID: "xmtp.org",
	typeID: "editMessage",
	versionMajor: 1,
	versionMinor: 0
)

/// Represents a request to edit a message.
/// This content type is used to request an edit of a specific message in a conversation.
/// Only the original sender can edit their own messages.
public struct EditMessageRequest: Codable, Equatable {
	public var messageId: String
	public var editedContent: EncodedContent?

	public init(messageId: String, editedContent: EncodedContent? = nil) {
		self.messageId = messageId
		self.editedContent = editedContent
	}
}

public struct EditMessageCodec: ContentCodec {
	public typealias T = EditMessageRequest

	public init() {}

	public var contentType: ContentTypeID = ContentTypeEditMessageRequest

	public func encode(content: EditMessageRequest) throws -> EncodedContent {
		let ffiEditedContent: FfiEncodedContent? = try content.editedContent.map { encoded in
			FfiEncodedContent(
				typeId: encoded.hasType
					? FfiContentTypeId(
						authorityId: encoded.type.authorityID,
						typeId: encoded.type.typeID,
						versionMajor: UInt32(encoded.type.versionMajor),
						versionMinor: UInt32(encoded.type.versionMinor)
					)
					: nil,
				parameters: encoded.parameters,
				fallback: encoded.hasFallback ? encoded.fallback : nil,
				compression: encoded.hasCompression ? Int32(encoded.compression.rawValue) : nil,
				content: encoded.content
			)
		}

		let ffi = FfiEditMessage(
			messageId: content.messageId,
			editedContent: ffiEditedContent
		)
		return try EncodedContent(serializedBytes: encodeEditMessage(request: ffi))
	}

	public func decode(content: EncodedContent) throws -> EditMessageRequest {
		let decoded = try decodeEditMessage(bytes: content.serializedBytes())
		return EditMessageRequest(
			messageId: decoded.messageId,
			editedContent: decoded.editedContent.map { ffiContent in
				var encoded = EncodedContent()
				if let type = ffiContent.typeId {
					var contentType = ContentTypeID()
					contentType.authorityID = type.authorityId
					contentType.typeID = type.typeId
					contentType.versionMajor = UInt32(type.versionMajor)
					contentType.versionMinor = UInt32(type.versionMinor)
					encoded.type = contentType
				}
				encoded.parameters = ffiContent.parameters
				if let fallback = ffiContent.fallback {
					encoded.fallback = fallback
				}
				encoded.content = Data(ffiContent.content)
				return encoded
			}
		)
	}

	public func fallback(content _: EditMessageRequest) throws -> String? {
		nil
	}

	public func shouldPush(content _: EditMessageRequest) throws -> Bool {
		false
	}
}

File: sdks/android/library/src/main/java/org/xmtp/android/library/codecs/EditMessageCodec.kt
package org.xmtp.android.library.codecs

data class EditMessageRequest(
    val messageId: String,
    val editedContent: EncodedContent?,
)

val ContentTypeEditMessageRequest =
    ContentTypeIdBuilder.builderFromAuthorityId(
        "xmtp.org",
        "editMessage",
        versionMajor = 1,
        versionMinor = 0,
    )

data class EditMessageCodec(
    override var contentType: ContentTypeId = ContentTypeEditMessageRequest,
) : ContentCodec<EditMessageRequest> {
    override fun encode(content: EditMessageRequest): EncodedContent {
        val ffiEditedContent =
            content.editedContent?.let { encoded ->
                uniffi.xmtpv3.FfiEncodedContent(
                    typeId =
                        encoded.type?.let { type ->
                            uniffi.xmtpv3.FfiContentTypeId(
                                authorityId = type.authorityId,
                                typeId = type.typeId,
                                versionMajor = type.versionMajor.toUInt(),
                                versionMinor = type.versionMinor.toUInt(),
                            )
                        },
                    parameters = encoded.parametersMap,
                    fallback = encoded.fallback.ifEmpty { null },
                    compression = encoded.compression?.let { it.number },
                    content = encoded.content.toByteArray(),
                )
            }

        val ffi =
            uniffi.xmtpv3.FfiEditMessage(
                messageId = content.messageId,
                editedContent = ffiEditedContent,
            )

        return EncodedContent.parseFrom(
            uniffi.xmtpv3.encodeEditMessage(ffi),
        )
    }

    override fun decode(content: EncodedContent): EditMessageRequest {
        val decoded = uniffi.xmtpv3.decodeEditMessage(content.toByteArray())

        val editedContent =
            decoded.editedContent?.let { ffiContent ->
                EncodedContent
                    .newBuilder()
                    .apply {
                        ffiContent.typeId?.let { type ->
                            setType(
                                ContentTypeId
                                    .newBuilder()
                                    .setAuthorityId(type.authorityId)
                                    .setTypeId(type.typeId)
                                    .setVersionMajor(type.versionMajor.toInt())
                                    .setVersionMinor(type.versionMinor.toInt())
                                    .build(),
                            )
                        }
                        putAllParameters(ffiContent.parameters)
                        ffiContent.fallback?.let { setFallback(it) }
                        ffiContent.compression?.let { setCompression(Compression.forNumber(it)) }
                        setContent(
                            com.google.protobuf.ByteString
                                .copyFrom(ffiContent.content),
                        )
                    }.build()
            }

        return EditMessageRequest(
            messageId = decoded.messageId,
            editedContent = editedContent,
        )
    }

    override fun fallback(content: EditMessageRequest): String? = null

    override fun shouldPush(content: EditMessageRequest): Boolean = false
}
